shell: Fix uart_rx_handle byte processing when ring buffer full

de-nordic opened the pull request:
In case when shell ring buffer gets full the data may still be taken
from the UART fifo, byte by byte, and accepted by the SMP back-end,
if the back-end has been enabled.

Unfortunately the uart_rx_handle failed to check if it has been
successfull in reading a byte from the fifo, before passing it to
the SMP for processing, which could lead to processing random stack data
as a part of an SMP frame.

Additinally, when the SMP would have accepted the byte,
the uart_rx_handle would fail to signal shell_thread, that the SMP has
internally buffered data that could be processed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
shell: Fix uart_rx_handle byte processing when ring buffer full

In case when shell ring buffer gets full the data may still be taken
from the UART fifo, byte by byte, and accepted by the SMP back-end,
if the back-end has been enabled.

Unfortunately the uart_rx_handle failed to check if it has been
successfull in reading a byte from the fifo, before passing it to
the SMP for processing, which could lead to processing random stack data
as a part of an SMP frame.

Additinally, when the SMP would have accepted the byte,
the uart_rx_handle would fail to signal shell_thread, that the SMP has
internally buffered data that could be processed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mniestroj commented on subsys/shell/shell_uart.c:
```diff
@@ -81,7 +81,9 @@ static void uart_rx_handle(struct device *dev,
 			 * is part of an mcumgr frame.
 			 */
-			smp_shell_rx_byte(&sh_uart->ctrl_blk->smp, dummy);
+			if (smp_shell_rx_byte(&sh_uart->ctrl_blk->smp, dummy)) {
```
This fix seems fine. However who said we have actually read some byte with:
```
rd_len = uart_fifo_read(dev, &dummy, 1);
```
?

Shouldn't we check `rd_len`?

de-nordic replied:
Yeah, this is actually good point and I have missed this one. I have been preparing fix for the SMP part which also fails to check `rd_len` and always sets `new_data` to `true`. Adding the fix here as second commit.

mniestroj replied:
I've seen your fix. Please also add to the commit message rationale why `rd_len != 0` was added and what it solves (prevents feeding smp_shell with random byte from stack).

de-nordic committed:
shell: Fix signal shell_thread only when data received

In case when SMP has been enabled, the shell_thread would be signaled
even if no data has been read from fifo.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
@mniestroj Thank you for comments and review!

nvlsianpu reviewed (approved):
I merged it in my head with #27657  
LGTM

de-nordic commented:
> 
> 
> I merged it in my head with #27657
> LGTM

@nvlsianpu Thanks for reviewing!

jakub-uC commented on subsys/shell/shell_uart.c:
```diff
@@ -78,10 +82,13 @@ static void uart_rx_handle(struct device *dev,
+			 * feeding it to SMP as a part of mcumgr frame.
 			 */
-			smp_shell_rx_byte(&sh_uart->ctrl_blk->smp, dummy);
+			if (rd_len != 0 &&
```
I know this is not compulsory but would you mind putting this condition within parentheses?
